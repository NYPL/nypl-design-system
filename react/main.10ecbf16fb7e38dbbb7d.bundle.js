(window.webpackJsonp=window.webpackJsonp||[]).push([[0],{1292:function(n,o,c){"use strict";c.r(o);c(1293),c(1323),c(1335),c(1442),c(1485)},1448:function(n,o){},1449:function(n,o){},1450:function(n,o){},1451:function(n,o){},501:function(n,o,c){c(502),c(645),n.exports=c(1292)},564:function(n,o){}},[[501,1,2]]]);