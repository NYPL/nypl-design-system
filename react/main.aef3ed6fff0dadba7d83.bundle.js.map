{"version":3,"file":"main.aef3ed6fff0dadba7d83.bundle.js","sources":["webpack:///main.aef3ed6fff0dadba7d83.bundle.js"],"mappings":"AAAA","sourceRoot":""}