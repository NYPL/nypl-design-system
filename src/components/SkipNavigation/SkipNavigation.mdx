import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";
import ComponentChangelogTable from "../../utils/ComponentChangelogTable";
import { changelogData } from "./skipNavigationChangelogData";

import * as SkipNavigationStories from "./SkipNavigation.stories";
import Link from "../Link/Link";

<Meta of={SkipNavigationStories} />

# SkipNavigation

| Component Version | DS Version   |
| ----------------- | ------------ |
| Added             | `0.28.0`     |
| Latest            | `Prerelease` |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#changelog" target="_self">Changelog</Link>}

## Overview

<Description of={SkipNavigationStories} />

**NOTE: In order to see the links in the `SkipNavigation` component, start by focusing
on the button below by tabbing or clicking on it. Then press "tab" to see the
first link and press "tab" again to see the second link. Hiding the links
intentional as the links should initially be visually hidden.**

## Component Props

<Canvas of={SkipNavigationStories.WithControls} />

<ArgTypes of={SkipNavigationStories.WithControls} />

## Accessibility

To create a useable page for screen readers, a link or a navigational list of
links should be added before all the content on a page. The Reservoir Design
System (DS) achieves this through the `SkipNavigation` component which renders
a `<nav>` element with an `aria-label` of "Skip Navigation". Inside the `<nav>`
element, an unordered list is rendered.

The reason this is a navigation list of links is because the first main link
points to the anchor of the main content, in this case `"#mainContent"`. The
second link points to accessibility information on [NYPL.org/accessibility](https://nypl.org/accessibility).

See the [Skip Navigation accessibility guide](../?path=/docs/accessibility-guide-skip-navigation--docs)
for more information on how to use this component with other DS components.

Resources:

- [WebAIM Skip Navigation Links](https://webaim.org/techniques/skipnav/)
- [WebAIM CSS in Action "Skip" links](https://webaim.org/techniques/css/invisiblecontent/#skipnavlinks)
- [CSS-Tricks A Deep Dive on Skipping to Content](https://css-tricks.com/a-deep-dive-on-skipping-to-content/)

## Changelog

<ComponentChangelogTable changelogData={changelogData} />
