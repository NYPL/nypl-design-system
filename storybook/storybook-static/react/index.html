<!doctype html><html lang="en"><head><meta charset="utf-8"><title>Storybook</title><meta name="viewport" content="width=device-width,initial-scale=1"><style>html, body {
    overflow: hidden;
    height: 100%;
    width: 100%;
    margin: 0;
    padding: 0;
  }</style><script>/* globals window */
  /* eslint-disable no-underscore-dangle */
  try {
    if (window.top !== window) {
      window.__REACT_DEVTOOLS_GLOBAL_HOOK__ = window.top.__REACT_DEVTOOLS_GLOBAL_HOOK__;
    }
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('unable to connect to top frame for connecting dev tools');
  }</script><style>#root[hidden],
      #docs-root[hidden] {
        display: none !important;
      }</style></head><body><div id="root"></div><div id="docs-root"></div><script>window['DOCS_MODE'] = false;</script><script src="runtime~main.00bd9d033d7a0b972fc9.bundle.js"></script><script src="vendors~main.d2b6aafef1a259d165c0.bundle.js"></script><script src="main.3bf71672a6dd35499004.bundle.js"></script></body></html>