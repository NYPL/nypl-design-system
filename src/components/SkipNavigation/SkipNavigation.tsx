import { Box, chakra, useStyleConfig } from "@chakra-ui/react";
import * as React from "react";

import Link from "../Link/Link";
import List from "../List/List";
import { ListTypes } from "../List/ListTypes";

interface SkipNavigationProps {
  /** Additional CSS class name to render in the `nav` element. */
  className?: string;
  /** ID that other components can cross reference for accessibility purposes */
  id?: string;
  /** The anchor target for the main skip link. The default is "#mainContent". */
  target?: string;
}

/**
 * SkipNavigation is a component that is used to provide a navigational list of
 * links. The first link is used to skip to the main content of the page using
 * the `#mainContent` id, and the second link points to accessibility information
 * on NYPL.org. These links are visually hidden but can be read by screenreaders.
 */
export const SkipNavigation = chakra(
  (props: React.PropsWithChildren<SkipNavigationProps>) => {
    const { className, id, target = "#mainContent", ...rest } = props;
    const styles = useStyleConfig("SkipNavigation");

    return (
      <Box
        as="nav"
        aria-label="Skip Navigation"
        className={className}
        id={id}
        __css={styles}
        {...rest}
      >
        <List inline noStyling type={ListTypes.Unordered}>
          <li>
            <Link href={target}>Skip to Main Content</Link>
          </li>
          <li>
            <Link href="https://www.nypl.org/accessibility">
              Click to learn about accessibility at the Library
            </Link>
          </li>
        </List>
      </Box>
    );
  }
);

export default SkipNavigation;
