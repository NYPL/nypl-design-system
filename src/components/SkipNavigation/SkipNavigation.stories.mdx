import { Box, VStack } from "@chakra-ui/react";
import {
  ArgsTable,
  Canvas,
  Description,
  Meta,
  Story,
} from "@storybook/addon-docs";

import Button from "../Button/Button";
import SkipNavigation from "./SkipNavigation";
import { getCategory } from "../../utils/componentCategories";

<Meta
  title={getCategory("SkipNavigation")}
  component={SkipNavigation}
  parameters={{
    jest: ["SkipNavigation.test.tsx"],
  }}
  argTypes={{
    className: { control: false },
    id: { control: false },
    target: { control: false },
  }}
/>

# SkipNavigation

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `1.0.0`    |
| Latest            | `1.0.0`    |

## Table of Contents

- [Overview](#overview)
- [Component Props](#component-props)
- [Accessibility](#accessibility)

## Overview

<Description of={SkipNavigation} />

**NOTE: In order to see the links in the `SkipNavigation` component, start by focusing
on the button below by tabbing or clicking on it. Then press "tab" to see the
first link and press "tab" again to see the second link. Hiding the links
intentional as the links should initially be visually hidden.**

## Component Props

<Canvas>
  <Story
    name="SkipNavigation with Controls"
    args={{
      className: undefined,
      id: "skip-nav",
      target: "mainContent",
    }}
  >
    {(args) => (
      <>
        <Button>Focus on me and then press "tab"</Button>
        <SkipNavigation {...args} />
      </>
    )}
  </Story>
</Canvas>

<ArgsTable story="SkipNavigation with Controls" />

## Accessibility

To create a useable page for screen readers, a link or a navigational list of
links should be added before all the content on a page. The Reservoir Design
System (DS) achieves this through the `SkipNavigation` component which renders
a `<nav>` element with an `aria-label` of "Skip Navigation". Inside the `<nav>`
element, an unordered list is rendered.

The reason this is a navigation list of links is because the first main link
points to the anchor of the main content, in this case `"#mainContent"`. The
second link points to accessibility information on [NYPL.org/accessibility](https://nypl.org/accessibility).

See the [Skip Navigation accessibility guide](https://nypl.github.io/nypl-design-system/storybook-static/?path=/story/accessibility-guide-skip-navigation--page)
for more information on how to use this component with other DS components.

Resources:

- [WebAIM Skip Navigation Links](https://webaim.org/techniques/skipnav/)
- [WebAIM CSS in Action "Skip" links](https://webaim.org/techniques/css/invisiblecontent/#skipnavlinks)
- [CSS-Tricks A Deep Dive on Skipping to Content](https://css-tricks.com/a-deep-dive-on-skipping-to-content/)
