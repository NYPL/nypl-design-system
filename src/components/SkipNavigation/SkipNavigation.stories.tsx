import type { Meta, StoryObj } from "@storybook/react";
import { withDesign } from "storybook-addon-designs";

import Button from "../Button/Button";
import SkipNavigation from "./SkipNavigation";

const meta: Meta<typeof SkipNavigation> = {
  title: "Components/Navigation/SkipNavigation",
  component: SkipNavigation,
  decorators: [withDesign],
  argTypes: {
    className: { control: false },
    id: { control: false },
    target: { control: false },
  },
};

export default meta;
type Story = StoryObj<typeof SkipNavigation>;

/**
 * Main Story for the SkipNavigation component. This must contains the `args`
 * and `parameters` properties in this object.
 */
export const Controls: Story = {
  args: {
    className: undefined,
    id: "skip-nav",
    target: "#mainContent",
  },
  render: (args) => (
    <>
      <Button id="example-btn">Focus on me and then press "tab"</Button>
      <SkipNavigation {...args} />
    </>
  ),
  parameters: {
    jest: ["SkipNavigation.test.tsx"],
  },
};
