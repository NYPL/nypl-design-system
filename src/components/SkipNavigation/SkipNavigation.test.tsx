import * as React from "react";
import { render, screen, within } from "@testing-library/react";
import { axe } from "jest-axe";
import renderer from "react-test-renderer";

import SkipNavigation from "./SkipNavigation";

describe("SkipNavigation Accessibility", () => {
  it("passes axe accessibility test", async () => {
    const { container } = render(<SkipNavigation />);
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("Label", () => {
  it("renders a nav element with an unordered list", () => {
    render(<SkipNavigation />);
    const nav = screen.getByRole("navigation");

    expect(nav).toBeInTheDocument();
    expect(nav).toHaveAttribute("aria-label", "Skip Navigation");
    // Chakra renders other lists so we only want the list we
    // created within the nav element.
    const { getByRole } = within(nav);
    expect(getByRole("list")).toBeInTheDocument();
  });

  it("renders two links", () => {
    render(<SkipNavigation />);
    const links = screen.getAllByRole("link");

    expect(links[0]).toHaveAttribute("href", "#mainContent");
    expect(links[1]).toHaveAttribute(
      "href",
      "https://www.nypl.org/accessibility"
    );
  });

  it("renders a custom main link target", () => {
    render(<SkipNavigation target="customId" />);
    const links = screen.getAllByRole("link");

    expect(links[0]).toHaveAttribute("href", "customId");
  });

  it("Renders the UI snapshot correctly", () => {
    const simple = renderer.create(<SkipNavigation />).toJSON();
    const customTarget = renderer
      .create(<SkipNavigation target="customId" />)
      .toJSON();
    const withChakraProps = renderer
      .create(<SkipNavigation p="20px" color="ui.error.primary" />)
      .toJSON();
    const withOtherProps = renderer
      .create(<SkipNavigation data-testid="props" />)
      .toJSON();

    expect(simple).toMatchSnapshot();
    expect(customTarget).toMatchSnapshot();
    expect(withChakraProps).toMatchSnapshot();
    expect(withOtherProps).toMatchSnapshot();
  });
});
