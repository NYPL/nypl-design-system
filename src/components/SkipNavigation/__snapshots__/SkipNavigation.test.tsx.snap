// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label Renders the UI snapshot correctly 1`] = `
<nav
  aria-label="Skip Navigation"
  className="css-1xdhyk6"
>
  <ul
    className="css-1xdhyk6"
  >
    <li>
      <a
        className="css-1xdhyk6"
        href="#mainContent"
        rel={null}
        target={null}
      >
        Skip to Main Content
      </a>
    </li>
    <li>
      <a
        className="css-1xdhyk6"
        href="https://www.nypl.org/accessibility"
        rel={null}
        target={null}
      >
        Click to learn about accessibility at the Library
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Label Renders the UI snapshot correctly 2`] = `
<nav
  aria-label="Skip Navigation"
  className="css-1xdhyk6"
>
  <ul
    className="css-1xdhyk6"
  >
    <li>
      <a
        className="css-1xdhyk6"
        href="customId"
        rel={null}
        target={null}
      >
        Skip to Main Content
      </a>
    </li>
    <li>
      <a
        className="css-1xdhyk6"
        href="https://www.nypl.org/accessibility"
        rel={null}
        target={null}
      >
        Click to learn about accessibility at the Library
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Label Renders the UI snapshot correctly 3`] = `
<nav
  aria-label="Skip Navigation"
  className="css-kle7zy"
>
  <ul
    className="css-1xdhyk6"
  >
    <li>
      <a
        className="css-1xdhyk6"
        href="#mainContent"
        rel={null}
        target={null}
      >
        Skip to Main Content
      </a>
    </li>
    <li>
      <a
        className="css-1xdhyk6"
        href="https://www.nypl.org/accessibility"
        rel={null}
        target={null}
      >
        Click to learn about accessibility at the Library
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Label Renders the UI snapshot correctly 4`] = `
<nav
  aria-label="Skip Navigation"
  className="css-1xdhyk6"
  data-testid="props"
>
  <ul
    className="css-1xdhyk6"
  >
    <li>
      <a
        className="css-1xdhyk6"
        href="#mainContent"
        rel={null}
        target={null}
      >
        Skip to Main Content
      </a>
    </li>
    <li>
      <a
        className="css-1xdhyk6"
        href="https://www.nypl.org/accessibility"
        rel={null}
        target={null}
      >
        Click to learn about accessibility at the Library
      </a>
    </li>
  </ul>
</nav>
`;
